mmv1/third_party/terraform/framework_utils/framework_utils.go

```diff
@@ -79,6 +79,47 @@ func getProjectFromFrameworkSchema(projectSchemaField string, rVal, pVal types.S
 	return types.String{}
 }
 
+func getRegionFramework(rRegion, rZone, pRegion, pZone types.String, diags *diag.Diagnostics) types.String {
+	return getRegionFromFrameworkSchema("region", rRegion, rZone, pRegion, pZone, diags)
+}
+
+// Infers the region based on the following (in order of priority):
+// - `regionSchemaField` in resource schema
+// - region extracted from the `zoneSchemaField` in resource schema
+// - provider-level region
+// - region extracted from the provider-level zone
+func getRegionFromFrameworkSchema(regionSchemaField string, rRegion, rZone, pRegion, pZone types.String, diags *diag.Diagnostics) types.String {
```

Comment:
I'm not a fan of having to pass in all these values, _but_ it's good to move away from the old pattern where we threw in the entirety of `*schema.ResourceData` as an argument.

Maybe we could have a struct that contains the relevant info only? 

Also, the `regionSchemaField` field here is doing nothing, except matching patterns from the old SDK version of the getRegion* functions. I think that made sense in the context of the SDK because everything was `map[string]interface{}` and you needed to know the key to use to access the correct state? But I don't thing it fulfils a purpose here (the error message could be hardcoded).

Reply:
In [8c5ace5](https://github.com/GoogleCloudPlatform/magic-modules/pull/7574/commits/8c5ace57bd0c1362eea76b4ce3b3fddb48b92db6) I've addressed this a bit by adding custom types that are just aliases for `types.String`

Reply:
The latest commit tries out a different approach for parameters in this function. Description in the team Slack channel! https://github.com/GoogleCloudPlatform/magic-modules/commit/347027eadbccad3a00ae95b529e2f92e180b3719